sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -417,18 +413,12 @@ class CodeGenContext {
       val code = expr.gen(this)
       val fn =
         s"""
-           |private void $fnName(InternalRow ${INPUT_ROW}) {
-           |  if (!$isLoaded) {
-           |    ${code.code.trim}
-           |    $isLoaded = true;
-           |    $isNull = ${code.isNull};
-           |    $value = ${code.value};
-           |  }
+           |private ${javaType(expr.dataType)} $fnName(InternalRow ${INPUT_ROW}) {
+           |  ${code.code.trim}
+           |  $isNull = ${code.isNull};
+           |  return ${code.value};
```

Comment:
and then we can remove the `subExprInitVariables` because the reset is just calling this function.